### PR TITLE
Speedup of pre-push hook

### PR DESCRIPTION
**Pull request by hoytak**
This PR speeds up the pre-push hook significantly, cutting down the pre-push hook execution time on the Flickr30k dataset, with one change committed, from 9.5 seconds to 5 seconds.   There are two main parts:

1. Initializing the CAS once, and using it for all operations.   The previous version create a CAS staging client three independent times.  Each time takes well over a second due to the time required to scan the cache folder (another issue, the next one to look at).

2. Running the operations in parallel while respecting the dependencies.  The uploading to the CAS is done in the background for all of it.  Uploading the shards is done in the background until the notes are needed.

### CODE REVIEW

File: rust/gitxetcore/src/errors.rs [hunk 1]
```diff
@@ -162,6 +165,7 @@ impl From<GitXetRepoError> for ExitCode {
             GitXetRepoError::RepoUninitialized(_) => 30,
             GitXetRepoError::RepoSaltUnavailable(_) => 31,
             GitXetRepoError::LazyConfigError(_) => 30,
```
[the remark this answers is not in the document]
Comment (hoytak) [reply rust/gitxetcore/src/errors.rs:167]: Hmmm.  Good point.  Separate PR?

File: rust/gitxetcore/src/merkledb_shard_plumb.rs [hunk 2]
```diff
@@ -606,7 +619,9 @@ pub async fn sync_session_shards_to_remote(
             git_xet_version: crate::data_processing_v2::GIT_XET_VERION.to_string(),
         };
 
-        let shard_file_client = shard_client::from_config(shard_connection_config).await?;
+        let shard_file_client = shard_client::from_config(shard_connection_config)
+            .await
+            .expect("1");
```
Comment (seanses) [inline rust/gitxetcore/src/merkledb_shard_plumb.rs:624]: Why `expect("1")` here?

File: rust/gitxetcore/src/merkledb_shard_plumb.rs [hunk 3]
```diff
@@ -620,17 +635,16 @@ pub async fn sync_session_shards_to_remote(
                 "Uploading shard {shard_prefix_ref}/{:?} from staging area to CAS.",
                 &si.shard_hash
             );
-            let data = fs::read(&si.path)?;
+            let data = fs::read(&si.path).expect("2");
```
Comment (seanses) [inline rust/gitxetcore/src/merkledb_shard_plumb.rs:638]: And `expect("2")` here?

File: rust/gitxetcore/src/merkledb_shard_plumb.rs (hunk 2 above)
Comment (hoytak) [reply rust/gitxetcore/src/merkledb_shard_plumb.rs:624]: Ah, was debugging.  good catch.

File: rust/gitxetcore/src/merkledb_shard_plumb.rs [hunk 4]
```diff
@@ -549,16 +550,12 @@ fn convert_merklememdb(
 
 /// Consolidate shards from sessions, install guard into ref notes v1 if
 /// a conversion was done. Write shards into ref notes v2.
-pub async fn sync_mdb_shards_to_git(
+pub async fn sync_session_mdb_shards_to_git(
```
Comment (seanses) [inline rust/gitxetcore/src/merkledb_shard_plumb.rs:553]: I prefer not to wrap `update_mdb_shards_to_git_notes` and `move_session_shards_to_local_cache` into a single function. `sync_session_mdb_shards_to_git` is only used once in this file and the name doesn't describe all the operations inside.

File: rust/gitxetcore/src/git_integration/git_xet_repo.rs [hunk 5]
```diff
@@ -1575,56 +1587,137 @@ impl GitXetRepo {
...
     }
 
     /// Pushes all the staged data in the local CAS
     pub async fn upload_all_staged(&self) -> Result<()> {
-        let repo = PointerFileTranslator::from_config(&self.xet_config).await?;
-        repo.upload_cas_staged(false).await?;
-        Ok(())
+        let cas = self.get_staging_cas().await?;
+
+        cas.upload_all_staged(MAX_CONCURRENT_UPLOADS, true)
```
Comment (seanses) [inline rust/gitxetcore/src/git_integration/git_xet_repo.rs:1625]: We don't want to retain uploaded staged cas blocks.

File: rust/gitxetcore/src/merkledb_shard_plumb.rs (hunk 4 above)
Comment (hoytak) [reply rust/gitxetcore/src/merkledb_shard_plumb.rs:553]: Agreed.  This was an intermediate function involved in how I was trying to split out the functionality into background and foreground stuff in the prepush hook.  Will fix.

File: rust/gitxetcore/src/git_integration/git_xet_repo.rs (hunk 5 above)
Comment (hoytak) [reply rust/gitxetcore/src/git_integration/git_xet_repo.rs:1625]: Ah.  Good catch.  Changed.

File: rust/gitxetcore/src/git_integration/git_xet_repo.rs [hunk 6]
```diff
@@ -1575,56 +1587,137 @@ impl GitXetRepo {
...
+
+                // Finally, we can move all the mdb shards from the session directory, which is used
+                // by the upload_shard task, to the cache.
+                move_session_shards_to_local_cache(
+                    &self.merkledb_v2_session_dir,
+                    &self.merkledb_v2_cache_dir,
+                )
+                .await?;
 
+                upload_all_jh.await??;
```
Comment (seanses) [inline rust/gitxetcore/src/git_integration/git_xet_repo.rs:1718]: Move this to before `sync_notes_to_remote` so there's no dependency issue in the remote?

Comment (hoytak) [reply rust/gitxetcore/src/git_integration/git_xet_repo.rs:1718]: In this case, the upload all task is just uploading all of the staged CAS blocks from the files. So it is independent of any of the notes.

**Review by seanses — APPROVED** [review]
LGTM!

**Pull request merged**